feat(iot-device): Add convention-based samples for IoT plug and play (#215)
Azure-Samples/azure-iot-samples-csharp: iot-hub/Samples/device/ConventionBasedOperations/readme.md

# IoT Plug And Play (PnP) device/module APIs

> Device(s)/module(s) connecting to IoT Hub that announce their DTDL model ID during initialization can now perform convention-based operations. One such convention supported is [IoT Plug and Play][pnp-convention].

Reviewer: ```suggestion
Device(s)/module(s) connecting to IoT Hub that announce their DTDL model ID during initialization can now perform convention-based operations. One such supported convention is [IoT Plug and Play][pnp-convention].
```

> These devices/modules can now use the native PnP APIs in the Azure IoT device SDKs to directly exchange messages with an IoT Hub, without having to manually format these messages to follow the PnP convention.

Reviewer: ```suggestion
These devices/modules can now use the Azure IoT Plug and Play APIs in the Azure IoT device SDKs to directly exchange messages with an IoT Hub, without having to manually format these messages to follow the Azure IoT Plug and Play convention.
```

}
```

#### Using convention-aware API (new):

```csharp
// Retrieve the client's properties.
 ClientProperties properties = await _deviceClient.GetClientPropertiesAsync(cancellationToken);

// To fetch the value of client reported property "serialNumber" under component "thermostat1".
bool isSerialNumberReported = properties.TryGetValue("thermostat1", "serialNumber", out string serialNumberReported);

// To fetch the value of service requested "targetTemperature" value under component "thermostat1".
bool isTargetTemperatureUpdateRequested = properties.Writable.TryGetValue("thermostat1", "targetTemperature", out double targetTemperatureUpdateRequest);
```

### Update top-level property:

#### Using non-convention-aware API (old):

```csharp
// Update the property "serialNumber".
var propertiesToBeUpdated = new TwinCollection
{
    ["serialNumber"] = "SR-1234",
};
await _deviceClient.UpdateReportedPropertiesAsync(propertiesToBeUpdated, cancellationToken);
```

#### Using convention-aware API (new):

```csharp
// Update the property "serialNumber".
var propertiesToBeUpdated = new ClientPropertyCollection
{
    ["serialNumber"] = "SR-1234",

> };

Reviewer: Does this still work with the AddRootProperty/AddComponentProperty changes?

> };

Reviewer: Do we want this to be part of the samples? It doesn't seem clear when we'd use `AddRootProperty` versus this.

Author: are you suggesting we use the explicit method call instead of an initializer list?

- Send telemetry.
- Update client properties and be notified of service requested property update requests.
- Respond to command invocation.

The samples demonstrate two scenarios:

- An IoT Plug and Play device that implements the [Thermostat][d-thermostat] model. This model has a single interface (the default component) that defines telemetry, properties and commands.
- An IoT Plug and Play device that implements the [Temperature controller][d-temperature-controller] model. This model defines multiple interfaces:
  - The top-level interface defines telemetry, properties and commands.
  - The model includes two [Thermostat][thermostat-model] components, and a [device information][d-device-info] component.

> NOTE: These samples are only meant to demonstrate the usage of Plug and Play APIs. If you are looking for a good device sample to get started with, please see the [device reconnection sample][device-reconnection-sample]. It shows how to connect a device, handle disconnect events, cases to handle when making calls, and when to re-initialize the `DeviceClient`.

> [pnp-convention]: https://docs.microsoft.com/azure/iot-pnp/concepts-convention

Reviewer: ```suggestion
[Azure IoT Plug and Play]: https://docs.microsoft.com/azure/iot-pnp/concepts-convention
```
